backend/apps/progress/services/progress_tracking_service.py

```
@@ -0,0 +1,197 @@
+    """
+    Domain service for handling complex user progress business logic.
+    Decoupled from HTTP request lifecycle and views.
+    """
+
+    @staticmethod
+    def record_lesson_progress(
+        user,
+        lesson_slug: str,
+        base_score: int = 100,
+        completed: bool = True,
+        idempotency_key: str = None,
```

Ruff (0.15.20)

[warning] PEP 484 prohibits implicit `Optional`

Convert to `T | None`

(RUF013)